### PR TITLE
Adding resource definition for IoTSecuritySolutionsAnalytics

### PR DESCRIPTION
**Pull request by hagba**
### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [x] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [x] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [ ] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

### CODE REVIEW

File: specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json [hunk 1]
```diff
@@ -0,0 +1,589 @@
...
+						}
+					}
+				},
+				"topAlertedDevices": {
+					"description": "The list of devices with the most attacked.",
+					"type": "array",
+					"items": {
+
+								"type": "#/definitions/IoTSecurityDeviceAlerts",
+								"description": "top 3 devices with most atacked alerts."
```
Comment (nschonni) [inline specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json]:
```suggestion
								"description": "top 3 devices with most attacked alerts."
```

File: specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json [hunk 2]
```diff
@@ -0,0 +1,589 @@
...
+				},
+				"description": {
+					"readOnly": true,
+					"type": "string",
+					"description": "Description of the incident and what it means"
+				},
+				"count": {
+					"readOnly": true,
+					"type": "int",
+					"description": "Occurance number of the alert within the aggregated date"
```
Comment (nschonni) [inline specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json]:
```suggestion
					"description": "Occurrence number of the alert within the aggregated date"
```

File: specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json [hunk 3]
```diff
@@ -0,0 +1,711 @@
+{
+  "swagger": "2.0",
+  "info": {
+    "title": "Security Center",
+    "description": "API spec for Microsoft.Security (Azure Security Center) resource provider",
+    "version": "2019-05-01-preview"
```
Comment (majastrz) [inline specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json]: >2019-05-01-preview [](start = 16, length = 18)

The version doesn't match the directory name. Which one should it be? `2017-08-01-preview` or `2019-05-01-preview`?

File: specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json [hunk 4]
```diff
@@ -0,0 +1,711 @@
...
+      "authorizationUrl": "https://login.microsoftonline.com/common/oauth2/authorize",
+      "flow": "implicit",
+      "description": "Azure Active Directory OAuth2 Flow",
+      "scopes": {
+        "user_impersonation": "impersonate your user account"
+      }
+    }
+  },
+  "paths": {
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Security/iotSecuritySolutions/{solutionName}/Analytics/current": {
```
Comment (majastrz) [inline specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json]: >current [](start = 147, length = 7)

For a singleton resource, the ARM naming convention is to call it `default`. Can you change the name?

Comment (majastrz) [reply specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json]: The resource type naming convention is camelCasePlural. Can you change the name to `analytics`?

---
In reply to: [295953066](https://github.com/Azure/azure-rest-api-specs/pull/6160#discussion_r295953066) [](ancestors = 295953066)

File: specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json [hunk 5]
```diff
@@ -0,0 +1,711 @@
...
+          "default": {
+            "description": "Error response describing why the operation failed.",
+            "schema": {
+              "$ref": "../../../common/v1/types.json#/definitions/CloudError"
+            }
+          }
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Security/iotSecuritySolutions/{solutionName}/Analytics/current/AggregatedAlert": {
```
Comment (majastrz) [inline specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json]: >AggregatedAlert [](start = 155, length = 15)

Can you rename to `aggregatedAlerts` to match the typical convention?

File: specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json [hunk 6]
```diff
@@ -0,0 +1,711 @@
...
+      "type": "oauth2",
+      "authorizationUrl": "https://login.microsoftonline.com/common/oauth2/authorize",
+      "flow": "implicit",
+      "description": "Azure Active Directory OAuth2 Flow",
+      "scopes": {
+        "user_impersonation": "impersonate your user account"
+      }
+    }
+  },
+  "paths": {
```
Comment (majastrz) [inline specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json]: >paths [](start = 3, length = 5)

You're missing the path to get all `analytics` resources. That needs to be implemented even if you just have 1. There are services in azure (Azure Policy is one) that scan and discover resources and they rely on these routes being implemented.

File: specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json [hunk 7]
```diff
@@ -0,0 +1,711 @@
...
+          "default": {
+            "description": "Error response describing why the operation failed.",
+            "schema": {
+              "$ref": "../../../common/v1/types.json#/definitions/CloudError"
+            }
+          }
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Security/iotSecuritySolutions/{solutionName}/Analytics/current/AggregatedAlert/{aggregatedAlertName}/Dismiss": {
```
Comment (majastrz) [inline specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json]: >Dismiss [](start = 193, length = 7)

Should be lower case `dismiss`

File: specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json [hunk 8]
```diff
@@ -0,0 +1,711 @@
...
+      "put": {
```
Comment (majastrz) [inline specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json]: >put [](start = 7, length = 3)

Actions on a resource are required to be POST as per the RPC. [https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/proxy-api-reference.md#resource-action-requests](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/proxy-api-reference.md#resource-action-requests). PUT is reserved for resource creation and update.

File: specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json [hunk 9]
```diff
@@ -0,0 +1,711 @@
...
+          "default": {
+            "description": "Error response describing why the operation failed.",
+            "schema": {
+              "$ref": "../../../common/v1/types.json#/definitions/CloudError"
+            }
+          }
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Security/iotSecuritySolutions/{solutionName}/Analytics/current/AggregatedRecommendation/{aggregatedRecommendationName}": {
```
Comment (majastrz) [inline specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json]: >AggregatedRecommendation [](start = 155, length = 24)

Can you rename to `aggregatedRecommendations`?

File: specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json (hunk 4 above)
Comment (majastrz) [reply specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json]: On 2nd thought analytics sounds really vague. I think analyticsModels or analyticModels would fit better and matches the name of the definition.

---
In reply to: [295954442](https://github.com/Azure/azure-rest-api-specs/pull/6160#discussion_r295954442) [](ancestors = 295954442,295953066)

File: specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json [hunk 10]
```diff
@@ -0,0 +1,711 @@
...
+        }
+      }
+    },
+    "IoTSecurityAggregatedAlertList": {
+      "description": "List of IoT aggregated security alerts",
+      "required": [
+        "value"
+      ],
+      "properties": {
+        "value": {
```
Comment (majastrz) [inline specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json]: >value [](start = 9, length = 5)

nit - you can make the value array read only also.

File: specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json [hunk 11]
```diff
@@ -0,0 +1,711 @@
...
+          "type": "integer",
+          "description": "count of medium severity items"
+        },
+        "low": {
+          "type": "integer",
+          "description": "count of low severity items"
+        }
+      }
+    },
+    "IoTSecuritySolutionAnalyticsModel": {
```
Comment (majastrz) [inline specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json]: >IoTSecuritySolutionAnalyticsModel [](start = 5, length = 33)

This doesn't look like an azure resource at all. The JSON should include properties like `id`, `name`, `type`, and `properties`. You should make this similar to `IoTSecurityAggregatedAlert` in the sense that you derive from a common resource definition and customize the contents of the `properties` property.

**Review by majastrz — CHANGES_REQUESTED** [review]
I added comments on the Swagger. Can you address the issues?

**Comment by hagba** [comment]
Thanks,
I will start working on them

File: specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json (hunk 3 above)
Comment (hagba) [reply specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json]: 2017-08-01-preview. fixed

File: specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json (hunk 4 above)
Comment (hagba) [reply specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json]: I understood that when I have only one model data I need to use current.
I have changed it to default

**Comment by hagba** [comment]
I have changed the files according to the comments and improved the example data.
I get errors in the automation tool that are not related to my changed files.

Thanks,
Hagit.

From: Hagit Badash
Sent: Thursday, June 20, 2019 11:24 PM
To: Azure/azure-rest-api-specs <reply@reply.github.com>; Azure/azure-rest-api-specs <azure-rest-api-specs@noreply.github.com>
Cc: Author <author@noreply.github.com>
Subject: RE: [Azure/azure-rest-api-specs] [Don't merge] Adding resource definition for IoTSecuritySolutionsAnalytics (#6160)

Thanks,
I will start working on them

**Review by ravbhatnagar — COMMENTED** [review]
some minor notes. please take a look.

File: specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json [hunk 12]
```diff
@@ -0,0 +1,757 @@
...
+        {
+          "$ref": "#/definitions/TagsResource"
+        }
+      ]
+    },
+    "IoTSecurityAggregatedAlertProperties": {
+      "type": "object",
+      "description": "Security Solution Aggregated Alert data",
+      "properties": {
+        "alertType": {
```
[the remark this answers is not in the document]
Comment (hagba) [reply specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json]: I do not want to limit it and update it on every alert that we will support in the future

File: specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json [hunk 13]
```diff
@@ -0,0 +1,757 @@
...
+          },
+          {
+            "$ref": "#/parameters/SolutionName"
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "OK",
+            "schema": {
+              "$ref": "#/definitions/IoTSecuritySolutionAnalyticsModel"
```
[the remark this answers is not in the document]
Comment (hagba) [reply specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json]: I want to emphasize that the list will always contain one value.

File: specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json [hunk 14]
```diff
@@ -0,0 +1,757 @@
...
+          "readOnly": true,
+          "type": "integer",
+          "description": "Occurrence number of the alert within the aggregated date"
+        },
+        "effectedResourceType": {
+          "readOnly": true,
+          "type": "string",
+          "description": "Azure resource ID of the resource that got the alerts"
+        },
+        "systemSource": {
```
[the remark this answers is not in the document]
Comment (hagba) [reply specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json]: It is not enum in the alert object in the alerts model (alerts.json)

File: specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json [hunk 15]
```diff
@@ -0,0 +1,777 @@
...
+      "description": "Security Solution Recommendation Information",
+      "properties": {
+        "recommendationName": {
+          "type": "string",
+          "description": "Name of the recommendation"
+        },
+        "recommendationDisplayName": {
+          "readOnly": true,
+          "type": "string",
+          "description": "Display name of the resommendation type"
```
Comment (nschonni) [inline specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json]:
```suggestion
          "description": "Display name of the recommendation type."
```

File: specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json (hunk 13 above)
[the remark this answers is not in the document]
Comment (ravbhatnagar) [reply specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/iotSecuritySolutionAnalytics.json]: yes, thats fine. but in a collection response, ARM expects a value array. You can have an array with one value.

**Comment by ravbhatnagar** [comment]
Looks good. Signing off from ARM side.

**Comment by jhendrixMSFT** [comment]
@AutorestCI regenerate azure-sdk-for-go

File: package.json [hunk 16]
```diff
@@ -50,5 +50,8 @@
     "oav": "oav",
     "tsc": "tsc",
     "multiapi": "ts-node ./scripts/multiapi.ts"
+  },
+  "dependencies": {
```
Comment (nschonni) [inline package.json]: Is this needed?

File: package.json [hunk 17]
```diff
@@ -50,5 +50,8 @@
...
+    "autorest": "^2.0.4283"
```
Comment (sergey-shandar) [inline package.json]: please, remove this line

**Review by sergey-shandar — APPROVED** [review]

**Comment by sergey-shandar** [comment]
@majastrz please, review the PR.

**Review by majastrz — APPROVED** [review]
Approving since Gaurav already marked it as ARMSignedOff.

**Pull request merged**